EventFilter/HcalRawToDigi/plugins/alpaka/HcalDigisProducerPortable.cc

```diff
@@ -0,0 +1,198 @@
+
+    using DeviceCollectionPhase1 = hcal::Phase1DigiDeviceCollection;
+    using DeviceCollectionPhase0 = hcal::Phase0DigiDeviceCollection;
+
+    // output product tokens
+    device::EDPutToken<DeviceCollectionPhase1> digisF01HEToken_;
+    device::EDPutToken<DeviceCollectionPhase0> digisF5HBToken_;
+    device::EDPutToken<DeviceCollectionPhase1> digisF3HBToken_;
+
+    struct ConfigParameters {
+      uint32_t maxChannelsF01HE, maxChannelsF5HB, maxChannelsF3HB;
+    };
+    ConfigParameters config_;
+  };
+
+  void HcalDigisProducerPortable::fillDescriptions(edm::ConfigurationDescriptions& confDesc) {
+    edm::ParameterSetDescription desc;
+
+    desc.add<edm::InputTag>("hbheDigisLabel", edm::InputTag("hcalDigis"));
+    desc.add<edm::InputTag>("qie11DigiLabel", edm::InputTag("hcalDigis"));
+    desc.add<std::string>("digisLabelF01HE", std::string{"f01HEDigisGPU"});
+    desc.add<std::string>("digisLabelF5HB", std::string{"f5HBDigisGPU"});
+    desc.add<std::string>("digisLabelF3HB", std::string{"f3HBDigisGPU"});
+    desc.add<uint32_t>("maxChannelsF01HE", 10000u);
+    desc.add<uint32_t>("maxChannelsF5HB", 10000u);
+    desc.add<uint32_t>("maxChannelsF3HB", 10000u);
+
+    confDesc.addWithDefaultLabel(desc);
+  }
+
+  HcalDigisProducerPortable::HcalDigisProducerPortable(const edm::ParameterSet& ps)
+      : hbheDigiToken_{consumes(ps.getParameter<edm::InputTag>("hbheDigisLabel"))},
+        qie11DigiToken_{consumes(ps.getParameter<edm::InputTag>("qie11DigiLabel"))},
+        digisF01HEToken_{produces(ps.getParameter<std::string>("digisLabelF01HE"))},
+        digisF5HBToken_{produces(ps.getParameter<std::string>("digisLabelF5HB"))},
+        digisF3HBToken_{produces(ps.getParameter<std::string>("digisLabelF3HB"))} {
+    config_.maxChannelsF01HE = ps.getParameter<uint32_t>("maxChannelsF01HE");
+    config_.maxChannelsF5HB = ps.getParameter<uint32_t>("maxChannelsF5HB");
+    config_.maxChannelsF3HB = ps.getParameter<uint32_t>("maxChannelsF3HB");
+  }
+
+  void HcalDigisProducerPortable::produce(device::Event& event, device::EventSetup const& setup) {
+    const auto& hbheDigis = event.get(hbheDigiToken_);
+    const auto& qie11Digis = event.get(qie11DigiToken_);
+
+    //Get the number of samples in data from the first digi
+    auto const stride = HBHEDataFrame::MAXSAMPLES * 0.5 + 1;
+    auto const size = hbheDigis.size() * stride;  // number of channels * stride
+
+    // stack host memory in the queue
+    HostCollectionPhase0 hf5_(size, event.queue());
+
+    // device product
+    DeviceCollectionPhase0 df5_(size, event.queue());
+
+    // set SoA_Scalar;
+    hf5_.view().stride() = stride;
+    hf5_.view().size() = hbheDigis.size();
+
+    for (unsigned int i = 0; i < hbheDigis.size(); ++i) {
+      auto const hbhe = hbheDigis[i];
+      auto const id = hbhe.id().rawId();
+      auto const presamples = hbhe.presamples();
+      uint16_t header_word = (1 << 15) | (0x5 << 12) | (0 << 10) | ((hbhe.sample(0).capid() & 0x3) << 8);
```

Comment:
is this incantation documented anywhere ?

Reply:
Unclear, this is copied from the CUDA implementation: https://github.com/cms-sw/cmssw/blob/master/EventFilter/HcalRawToDigi/plugins/HcalDigisProducerGPU.cc#L166

Reply:
I see... undocumented for undocumented, might as well leave it like it is.